src/shared/ToplevelMappingManager.cpp

```diff
@@ -0,0 +1,40 @@
+#include "ToplevelMappingManager.hpp"
+#include "../helpers/Log.hpp"
+#include "../core/PortalManager.hpp"
+
+CToplevelMappingManager::CToplevelMappingManager(SP<CCHyprlandToplevelMappingManagerV1> mgr) : m_pManager(mgr) {
+    Debug::log(LOG, "[toplevel mapping] registered manager");
+}
+
+void CToplevelMappingManager::fetchWindowForToplevel(SP<CCZwlrForeignToplevelHandleV1> handle) {
+    if (!handle)
+        return;
+
+    Debug::log(TRACE, "[toplevel mapping] fetching window for toplevel at {}", (void*)handle.get());
+    auto const HANDLE = makeShared<CCHyprlandToplevelWindowMappingHandleV1>(m_pManager->sendGetWindowForToplevelWlr(handle->resource()));
+
+    m_vHandles.push_back(HANDLE);
+
+    HANDLE->setWindowAddress([this, handle](CCHyprlandToplevelWindowMappingHandleV1* h, uint32_t address_hi, uint32_t address) {
+        const auto ADDRESS = (uint64_t)address_hi << 32 | address;
+        this->m_muAddresses.insert_or_assign(handle, ADDRESS);
+        Debug::log(TRACE, "[toplevel mapping] mapped toplevel at {} to window {}", (void*)handle.get(), ADDRESS);
+        std::erase_if(this->m_vHandles, [&](const auto& other) { return other.get() == h; });
+    });
+
+    HANDLE->setFailed([this, handle](CCHyprlandToplevelWindowMappingHandleV1* h) {
+        Debug::log(TRACE, "[toplevel mapping] failed to map toplevel at {} to window", (void*)handle.get());
+        std::erase_if(this->m_vHandles, [&](const auto& other) { return other.get() == h; });
```

Comment:
we dont use this->

Reply:
replaced it with `g_pPortalManager->m_sHelpers.toplevelMapping`

Reply:
you misunderstood. Capturing `this` is completely fine. I was just talking about `this->`. It's implicit, so you can just drop it